### ISSUE TITLE
Feature Request: Whitelisting characters

### ISSUE THREAD

**Issue by basic-bgnr** [robertknight/ocrs#114]
Great project by the way. Thank you.

I'm running into a problem using `ocrs` that a lot of false positive values are generated because of the way my images are structured (i.e the images doesn't contains digits but only regular english alphabet.). I'm getting a lot of 0 -> O, l->1, and s->5 .. errors. I would like to whitelist a set of characters just as tesseract-ocr allows to do. But I think this feature is missing.

**Comment by robertknight** [comment]
> I would like to whitelist a set of characters just as tesseract-ocr allows to do. But I think this feature is missing.

Indeed, there is currently no API for adjusting the probabilities of characters after the recognition model runs. Adding a whitelist would involve setting the probabilities of all excluded characters to zero ([here](https://github.com/robertknight/ocrs/blob/9a686feb519c474b7722b8ec2c0f23cbda25384e/ocrs/src/recognition.rs#L513)) before decoding.

**Comment by basic-bgnr** [comment]
@robertknight 
I was just going through the code as you pointed here https://github.com/robertknight/ocrs/blob/9a686feb519c474b7722b8ec2c0f23cbda25384e/ocrs/src/recognition.rs#L513 and modified it for a simple proof of concept. 

    let white_list = "abcdefghijklmnopqrstuvwxyz";
    let blacklist_index = alphabet
        .chars()
        .enumerate()
        .filter_map(|(index, char)| {
            if !white_list.contains(char) {
                Some(index)
            } else {
                None
            }
        })
        .collect::<Vec<_>>();
      
    for row in 0..input_seq.shape()[0] {
        for &column in blacklist_index.iter() {
            input_seq[[row, column]] = 0.0;
        }
    }

I've made needed mutability change in the other part of code to actually compile the lib but it isn't giving the correct output. Could you point me in correct direction?

**Comment by robertknight** [comment]
Can you post a few test images that show examples of where you want the output to favor eg. "0" over "O" or make other adjustments?

**Comment by basic-bgnr** [comment]
@robertknight Thank you. I managed to get it working for my use case.  I had to go through your model training repo  [here](https://github.com/robertknight/ocrs-models/blob/3d98fc655d6fd4acddc06e7f5d60a55b55748a48/ocrs_models/datasets/util.py#L113) to find out. 
I was encoding the character wrong, I had to ` + 1` the character index. A small comment on why `-1` was done [here](https://github.com/robertknight/ocrs/blob/main/ocrs/src/recognition.rs#L295) would be really helpful to other readers.  
And lastly I had to set  the probability of the excluded characters to `f32::NEG_INFINITY` instead of `0.0`. 

Would you like a PR on this?

**Comment by robertknight** [comment]
> A small comment on why -1 was done [here](https://github.com/robertknight/ocrs/blob/main/ocrs/src/recognition.rs#L295) would be really helpful to other readers.

I agree. Would you mind adding a comment if you're submitting a PR for this feature.

> Would you like a PR on this?

Sure, this seems like useful functionality. It would probably be a good idea to sketch out the API in this issue first. The simplest I can think of might be something like:

```rs
OcrEngineParams {
  // ...existing fields
  
  /// Characters from the recognition model's alphabet which will be predicted.
  ///
  /// Defaults to the recognition model's full alphabet.
  allowed_characters: Option<String>,
}
```

The API could be much more general (eg. allowing the user to specify a custom weight-processing function), but I'd be inclined to have a simple API for a more common use case like this.

Thoughts?

**Comment by basic-bgnr** [comment]
@robertknight 

> It would probably be a good idea to sketch out the API in this issue first

Well for my current implement, I haven't introduced any new APIs. I've modified the internal code similar in line to what you've thought. i.e
```
OcrEngineParams {
  // ...existing fields
  
  /// Characters from the recognition model's alphabet which will be predicted.
  ///
  /// Defaults to the recognition model's full alphabet.
  allowed_characters: Option<String>,
}
```
Frankly speaking I don't have any experience of API design so I don't think I would be of much help in this regards.
But If I had to put my opinion, right now, exposing public API for custom weight processing would be too much of work. 

If there's requirement (or feature request) of such general API in the future, this internal implementation could be replaced by using the API internally to make the code cleaner.

**Comment by robertknight** [comment]
The reason for the log-softmax activation is because PyTorch's [`CTCLoss`](https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html) expects log probabilities as inputs, as does CTC decoding in Ocrs. For the moment I would consider the use of log probability outputs from the model part as part of its "API contract".

I toyed with some design possibilities for an API, however something that became clear to me is that in order to use a flexible API, that would involve exposing a lot more of the implementation details to the end user - they have to learn a bit about CTC decoding etc. I would suggest to stick with something simple like I sketched out above, and the API can evolve in future when other use cases are more clear.

**Comment by basic-bgnr** [comment]
@robertknight
> For the moment I would consider the use of log probability outputs from the model part as part of its "API contract".
 
yeah that would be nice (explicit is better than implicit ....)  


>  I toyed with some design possibilities for an API, however something that became clear to me is that in order to use a flexible API, that would involve exposing a lot more of the implementation details to the end user - they have to learn a bit about CTC decoding etc.   

I mean, for people are using the default model, it would be expecting a lot but for lib user who are using custom-model, it wouldn't be that hard for them to know the internal details. But fleshing out a suitable API in itself is a hard job. 

>  I would suggest to stick with something simple like I sketched out above, and the API can evolve in future when other use cases are more clear.

ok nice. I've already worked out a solution in line to what you've suggested earlier, should a open a PR then ?

**Comment by robertknight** [comment]
> I've already worked out a solution in line to what you've suggested earlier, should a open a PR then ?

Please do 👍

**Comment by basic-bgnr** [comment]
thanks !.
